src/components/Hoverable/index.js

```diff
@@ -1,147 +1,101 @@
-            return;
-        }
-
-        if (this.wrapperView.contains(e.target)) {
-            return;
-        }
-
-        this.setIsHovered(false);
-    }
-
-    handleVisibilityChange() {
-        if (document.visibilityState !== 'hidden') {
-            return;
-        }
-
-        this.setIsHovered(false);
-    }
-
-    render() {
-        let child = this.props.children;
-        if (_.isArray(this.props.children) && this.props.children.length === 1) {
-            child = this.props.children[0];
-        }
-
-        if (_.isFunction(child)) {
-            child = child(this.state.isHovered);
-        }
-
-        if (!DeviceCapabilities.hasHoverSupport()) {
-            return child;
-        }
-
-        return React.cloneElement(React.Children.only(child), {
-            ref: (el) => {
-                this.wrapperView = el;
-
-                // Call the original ref, if any
-                const {ref} = child;
-                if (_.isFunction(ref)) {
-                    ref(el);
-                    return;
-                }
-
-                if (_.isObject(ref)) {
-                    ref.current = el;
-                }
-            },
-            onMouseEnter: (el) => {
-                this.setIsHovered(true);
-
-                if (_.isFunction(child.props.onMouseEnter)) {
-                    child.props.onMouseEnter(el);
-                }
-            },
-            onMouseLeave: (el) => {
-                this.setIsHovered(false);
-
-                if (_.isFunction(child.props.onMouseLeave)) {
-                    child.props.onMouseLeave(el);
-                }
-            },
-            onBlur: (el) => {
-                // Check if the blur event occurred due to clicking outside the element
-                // and the wrapperView contains the element that caused the blur and reset isHovered
-                if (!this.wrapperView.contains(el.target) && !this.wrapperView.contains(el.relatedTarget)) {
-                    this.setIsHovered(false);
-                }
-
-                if (_.isFunction(child.props.onBlur)) {
-                    child.props.onBlur(el);
-                }
-            },
-        });
-    }
+
+function Hoverable({disabled, onHoverIn, onHoverOut, children}) {
+    const [isHovered, setIsHovered] = useState(false);
+    const wrapperView = useRef(null);
+
+    useEffect(() => {
+        // For now, listener is quite simple so we don't need to extract its declaration from effect
+        const onVisibilityChange = () => document.visibilityState === 'hidden' && setIsHovered(false);
+
+        document.addEventListener('visibilitychange', onVisibilityChange);
+
+        return () => document.removeEventListener('visibilitychange', onVisibilityChange);
+    }, []);
+
+    useEffect(() => {
+        if (disabled) return setIsHovered(false);
+    }, [disabled]);
+
+    useEffect(() => {
+        if (isHovered) return onHoverIn();
```

Comment:
I think here is missing the check `if (disabled) { return; }` before doing any `onHoverXXX`

Reply:
That's a great point! I think we can discuss how flow with `disabled` could look like, as it is now conflicting with internal `isHovered` state declaration (for example we are not prohibiting event handlers to be executed even if `disabled` is set to `true`)

Reply:
Done ✅